### ISSUE TITLE
SuperCollider / Superdirt / Bluetooth headphones affecting speed of playback

### ISSUE THREAD

**Issue by Ouyuegu** [supercollider/supercollider#5729]
Hello all,

I'm not especially well-versed in the ways of SuperCollider, and only know how to use it in the most basic way, to use TidalCycles, so forgive me if I am asking any stupid questions.

For convenience sake, I sometimes want to be able to use my AirPods Pro to play about with Tidal, but am having a weird issue with it.

At first, when trying to launch Superdirt, I was told to set input channels to 0, due to the sample-rate mismatch, so I ran the following:

```supercollider
o = Server.default.options;
o.numInputBusChannels = 0
```

And that seemed to work, and I got connected, but now the sound running Tidal through Atom plays all samples and cycles at like 10% speed. Meaning it actually slows down the samples, as well as the cycle speed... I've never had this issue before in any context, so I wouldn't know where to begin with diagnosing it. 

I have tested it on my computer's internal speakers, and also through the headphone jack...  this only happens when I connect to my Apple AirPods Pros.

Any help you can give would be much appreciated.

Many, many thanks in advance!

**Comment by dyfer** [comment]
I believe bluetooth headphones may run at lower samplerate, which is either not being picked up by SuperCollider, or is otherwise mismatched (i.e. SC is producing output at 48k but audio device is only processing it at 16kHz or so).
What if you requested a particular samplerate, like `o.sampleRate = 48000` (or 44100) before booting the server? (or maybe try to actually run at 16k...?)
Finally, you can try creating an aggregate device (in Audio MIDI setup) with built-in speakers and airpods, set "drift correction" for airpods and built-in sp as clock master and use that as a device with SC (that way coreaudio backend should resample from the rate the aggregate device runs at to whatever airpods run at).

**Comment by telephon** [comment]
At least for me, checking the Audio MIDI Setup in macOS, the AirPods Pro have an 48 kHz / 32 bit float option for the output. The input is limited to 16 kHz 16 bit integer.

Because of this discrepancy, you need to set the `outDevice` separately:

```supercollider
s.options.outDevice = "wanninger’s AirPods P";
s.options.outDevice = "MacBook Pro Microphone";
```

**Comment by Ouyuegu** [comment]
Many thanks for your help! My problem is now sorted after following @dyfer 's suggestion to request the sample rate before booting the server.

@telephon Thanks for your advice, also!

**Comment by ethanfischer** [comment]
I was having a similar issue with my airpods pro on my macbook. Except that it was only playing things in mono, instead of stereo.

Here's what worked for me:

On mac, open Audio MIDI Setup. Add a new aggregate device. Click the checkbox for airpods pro and macbook pro microphone. For clock source, select airpods pro. Click the drop down and make sure both options are set to sample rate 48kHz. Make sure macbook pro microphone has drift correction selected. Rename this aggregate device "sc".

In supercollider, make sure your servers are killed and then run the following:

s.options.device = "sc";
s.boot;

You should see something like the following in the Post window:

Requested sample rate 44100.000000 was not available - attempting to use sample rate of 48000.000000
"sc" Input Device
   Streams: 1
      0  channels 1

"sc" Output Device
   Streams: 1
      0  channels 2

SC_AudioDriver: sample rate = 48000.000000, driver's block size = 512
SuperCollider 3 server ready.


The important part is that your ouput device is recognizing 2 channels.

Hope this helps someone!
